### ISSUE TITLE
How to write a Z5 metadata to enable a raw compression

### ISSUE THREAD

**Issue by halehawk** [constantinpape/z5#110]
I want to see the performance of writing zarr/Z5 file without compression. But I don't know how to set Z5 metadata to enable a raw compression, whatever I change on metadata, I get json error or metadata error. Could you please help me on it? Thanks. Haiying

**Comment by constantinpape** [comment]
Are you using the python API or the C++ API?
In python you need to set `compression='raw'` when calling `create_dataset` / `require_dataset`.
In C++, you need to set the argument `compressor` to raw when you use the [factory function](https://github.com/constantinpape/z5/blob/master/include/z5/dataset_factory.hxx#L61).

I could be of more help, if you post an example of what you're trying to do and the error you get.

**Comment by halehawk** [comment]
I ran your test_factories.cxx to test factory.  Or I added

meta.compressor = types::raw;

in the writeMetadata in the test_factories.cxx.


I always got the following errors:

Running main() from gtest_main.cc

[==========] Running 2 tests from 1 test case.

[----------] Global test environment set-up.

[----------] 2 tests from FactoryTest

[ RUN      ] FactoryTest.OpenAllDtypes

before compressorid

unknown file: Failure

C++ exception with description "[json.exception.type_error.305] cannot use
operator[] with a string argument with null" thrown in the test body.

[  FAILED  ] FactoryTest.OpenAllDtypes (2 ms)

[ RUN      ] FactoryTest.CreateAllDtypes

[       OK ] FactoryTest.CreateAllDtypes (9 ms)

[----------] 2 tests from FactoryTest (11 ms total)


[----------] Global test environment tear-down

[==========] 2 tests from 1 test case ran. (11 ms total)

[  PASSED  ] 1 test.

[  FAILED  ] 1 test, listed below:

[  FAILED  ] FactoryTest.OpenAllDtypes


 1 FAILED TEST


Only after I added

        meta.compressor = types::zlib;

        meta.compressionOptions["level"] = 5;

        meta.compressionOptions["useZlib"] = true;


in the writeMetadata in the test_factories.cxx.


I can get run passed.


Thank you very much for your help,


Haiying

**Comment by constantinpape** [comment]
Sorry for the late reply.

I couldn't find a way to reproduce this. Setting compression to `raw` in `test_factories.cxx` works for me.
See [here](https://github.com/constantinpape/z5/blob/beac76c7bcb3fd56cceed99d1e9b8eb6f11fc031/src/test/test_factories.cxx#L19).

**Comment by clbarnes** [comment]
In aid of making these sorts of options more discoverable, what would you think about creating an enum covering all the accepted values? Using enum34_custom.StrEnum you can make it interchangeable with strings (I think).

**Comment by halehawk** [comment]
Could you please tell me which version of nlohmann you are using now?

**Comment by halehawk** [comment]
Also, do you have benchmark results of Z5? I checked on your git repo and
didn't find it.

**Comment by constantinpape** [comment]
> Could you please tell me which version of nlohmann you are using now?

I just checked and I am using 3.5.0 (from conda-forge).

> Also, do you have benchmark results of Z5? I checked on your git repo and didn't find it.

I set up an [asv benchmark for z5](https://github.com/constantinpape/z5py-benchmarks) a while ago.
I also set up hosting of results via [github.io](https://constantinpape.github.io/z5py-benchmarks/#/).
Unfortunately something seems to be off and no results are displayed.
If you want you can open a separate issue about this in the benchmarks repo.

**Comment by constantinpape** [comment]
@clbarnes

> In aid of making these sorts of options more discoverable, what would you think about creating an enum covering all the accepted values?

These options are stored in enums already, see [here](https://github.com/constantinpape/z5/blob/master/include/z5/types/types.hxx#L76-L93). I am using a map to convert between the enum and strings, see [here](https://github.com/constantinpape/z5/blob/master/include/z5/types/types.hxx#L101-L122).



> Using enum34_custom.StrEnum you can make it interchangeable with strings (I think).

As far as I understand `enum34_custom` is a python module, so I don't see how this would help much on the c++ side. 

The general point that all this should be more discoverable is very true; there is no documentation for the c++ side of the project.

**Comment by constantinpape** [comment]
This seems to be resolved.
@halehawk let me know if there are still issues related to the initial issue.
